### DISCUSSION TITLE
Re-exporting pyo3 breaks macros

### DISCUSSION THREAD

**Discussion by GrinningSoulGH** [PyO3/pyo3#1978] [Questions] [answered]
I've been using a wrapper crate to sync versions of pyo3, pyo3-log and pyo3-asyncio between different lib and binary crates I'm developing. However, some of the code in macros appears to rely on pyo3 being available in the global scope (example: 21196e68e289322f84cfc3aa50c6cbd34e4985ca). This is not the case for me, as pyo3 comes as a dependency for the wrapper crate. Could you elaborate on why that choice was made?

**Comment by mejrs** [comment]
> However, some of the code in macros appears to rely on pyo3 being available in the global scope

That (series of) commit(s) made it so that pyo3 *doesn't* rely on the things in the user's scope. Can you clarify what you mean? 

Or are you talking about macros in pyo3-log/pyo3-asyncio? If so you should file an issue there, because IMO a macro being unhygienic is a bug.

**Comment by birkenfeld** [comment]
I guess the problem is that if `pyo3` is not a dependency of the current crate, the expanded code using `::pyo3::Foo` will not automagically find `pyo3` at the crate root.

However, previously `pyo3` was *still* expected somewhere (crate root or current scope)...

**Reply by GrinningSoulGH** [reply]
It's not in the crate root, as you ascertained. Well, I can add it explicitly, but it makes the versions drift on every update. The question really becomes whether I can make `pyo3` available in current scope under the required name `::pyo3` and not `::my_crate::pyo3`, as I _roughly_ have at the moment. And, also, whether this is an issue, as I haven't encountered that "scope" requirement with any of the crates I've used.

**Comment by davidhewitt** [comment, marked as answer]
I think this is one of those things where it hurts both ways around. 

As the other answers say, we felt it was better to anchor the macros to the global scope rather than relying on `pyo3` being defined correctly at in the current scope. This leads to fewer suprises where PyO3 (or std) can be overridden in the local scope with unexpected effects.

OTOH, I can certainly appreciate your use case wanting to write `use my_crate::pyo3` to override what the macros look at. I think a number of proc macro crates are in similar positions where it's not clear what's better on this front. I'm happy with what we currently do because it's more robust, however am not against reconsidering.

**Comment by birkenfeld** [comment]
Maybe serde's approach is where we should aim at: https://serde.rs/container-attrs.html#crate

Basically, allowing `#[pyo3(crate = "...")]` to specify where to find the pyo3 crate.

**Reply by davidhewitt** [reply]
This is a great idea!
